_posts/2017-09-08-scalafix-v0.5.md

```diff
@@ -0,0 +1,190 @@
+
+
+## Linting
+
+Scalafix v0.5 supports linters.
+Before v0.5, Scalafix rules could only "fix" code by providing rewrite instructions.
+Now, it's possible to implement Scalafix rules that report lint messages instead
+of rewrite instructions.
+Each category of lint messages is attached to a default severity level
+(info/warn/error), which users can override via configuration.
+
+The first Scalafix linter is [`NoInfer`][].
+`NoInfer` reports errors when the Scala compiler infers one of the following types:
+`Any`, `AnyVal`, `Product` or `Serializable`.
+It's usually a bad sign when the compiler infers such a generic type.
+In those cases, it's better to explicitly type annotate the expected type.
+
+
+### Building a custom linter
+It's simple to implement custom Scalafix linters.
+To demonstrate this, let's implement a `NoDefaultParams` linter that reports an error if
+default parameter values are used.
+To begin with, we motivate this linter rule with a bit of documentation.
+
+<script src="https://gist.github.com/olafurpg/38699f591329c3bcff52b7412dcb696b.js"></script>
+
+Looks good, now we can implement the Scalafix rule.
+Rules are implemented with Scalafix and Scalameta.
+The `param"..."` part is a [quasiquote][], which is a simple way to construct
+or pattern-match a Scalameta tree from syntax.
+
+<script src="https://gist.github.com/olafurpg/7b022c7120156716def282028b091e8b.js"></script>
+
+When `NoDefaultParams` encounters a parameter with a default parameter value,
+it will report a message like this
+
+```scala
+scala/test/NoDefaultParams.scala:8: error: [NoDefaultParams] default parameter values are not allowed.
+  def foo(e: Int = 2) = e
+                   ^
+```
+
+We can use Scalafix-testkit to test that our rewrite works as expected.
+Tests are written in plain .scala files.
+The top of the file contains a comment including configuration to tell Scalafix
+which rule to run.
+Rules can also have their own configuration section, but that's out of scope
+for this blog post.
+
+<script src="https://gist.github.com/olafurpg/be6b0fe5b317de7555dc0b6cd0aebfbf.js"></script>
+
+The `// assert: ` comment asserts that a linter message with id "NoDefaultParams" is reported
+on line 7.
+The test fails if
+
+- a message is not reported on that line
+- a message is reported on that line but the category id doesn't match,
+- a message is reported at a line without an assert
+
+`NoDefaultParams` have docs, an implementation and tests, our linting rule is
+ready for a release!
+Our friends can run this rule on their codebase by installing [sbt-scalafix][]
+and execute from the sbt shell
+
+```
+sbt> scalafix https://gist.githubusercontent.com/olafurpg/7b022c7120156716def282028b091e8b/raw/2a83cfc22462804c53a9fb67d0c0220b8b82f124/NoDefaultParameters.scala
+```
+
+Note that there is no need to publish the linter rule to Maven Central.
+Scalafix is able to run custom linting rules from source.
```

Comment:
I think what's really missing is the ability to run rules from adhoc classpaths. This seems to strike a perfect balance between a single file (easy to get started, hard to set up test harness) and Maven Central (hard to get started, easy to add custom logic to the project).

Reply:
Yes, that's a good feature idea.